src/envs/pybullet_cover.py

```diff
@@ -0,0 +1,313 @@
+            self._block_ids.append(
+                create_pybullet_block(color, half_extents, self._obj_mass,
+                                      self._obj_friction, orientation,
+                                      self._physics_client_id))
+        self._target_ids = []
+        for i in range(CFG.cover_num_targets):
+            color = self._obj_colors[i % len(self._obj_colors)]
+            color = (color[0], color[1], color[2], 0.5)  # slightly transparent
+            width = CFG.cover_target_widths[i] / max_width * self._max_obj_width
+            half_extents = (self._obj_len_hgt / 2.0, width / 2.0,
+                            self._target_height / 2.0)
+            orientation = [0.0, 0.0, 0.0, 1.0]  # default
+            self._target_ids.append(
+                create_pybullet_block(color, half_extents, self._obj_mass,
+                                      self._obj_friction, orientation,
+                                      self._physics_client_id))
+
+    def _create_pybullet_robot(
+            self, physics_client_id: int) -> _SingleArmPyBulletRobot:
+        ee_home = (self._workspace_x, self._robot_init_y, self._workspace_z)
+        return create_single_arm_pybullet_robot(
+            CFG.pybullet_robot, ee_home, self._ee_orn, self._open_fingers,
+            self._closed_fingers, self._move_to_pose_tol, self._max_vel_norm,
+            self._grasp_tol, physics_client_id)
+
+    def _extract_robot_state(self, state: State) -> Array:
+        if self._HandEmpty_holds(state, []):
+            fingers = self._open_fingers
+        else:
+            fingers = self._closed_fingers
+        y_norm = state.get(self._robot, "hand")
+        # De-normalize robot y to actual coordinates.
+        ry = self._y_lb + (self._y_ub - self._y_lb) * y_norm
+        rx = state.get(self._robot, "pose_x")
+        rz = state.get(self._robot, "pose_z")
+        return np.array([rx, ry, rz, fingers], dtype=np.float32)
+
+    def _reset_state(self, state: State) -> None:
+        """Run super(), then handle cover-specific resetting."""
+        super()._reset_state(state)
+
+        # Reset blocks based on the state.
+        block_objs = state.get_objects(self._block_type)
+        self._block_id_to_block = {}
+        for i, block_obj in enumerate(block_objs):
+            block_id = self._block_ids[i]
+            self._block_id_to_block[block_id] = block_obj
+            bx = self._workspace_x
+            # De-normalize block y to actual coordinates.
+            y_norm = state.get(block_obj, "pose")
+            by = self._y_lb + (self._y_ub - self._y_lb) * y_norm
+            if state.get(block_obj, "grasp") != -1:
+                # If an object starts out held, it has a different z.
+                bz = self._workspace_z - 0.01
+            else:
+                bz = self._table_height + self._obj_len_hgt * 0.5
+            p.resetBasePositionAndOrientation(
+                block_id, [bx, by, bz], [0.0, 0.0, 0.0, 1.0],
```

Comment:
this default block orientation is used in enough places that we should make it a class variable